lib/cfa/augeas_parser/writer.rb

```diff
@@ -0,0 +1,289 @@
+        process_operation(located_entry)
+      end
+      @lazy_operations.run if top_level
+    end
+
+  private
+
+    # AugeasEntry togethere with information about its location and few
+    # helper methods to detect siblings.
+    class LocatedEntry
+      attr_reader :prefix
+      attr_reader :entry
+      attr_reader :tree
+
+      def initialize(tree, entry, prefix)
+        @tree = tree
+        @entry = entry
+        @prefix = prefix
+      end
+
+      def orig_key
+        entry[:orig_key]
+      end
+
+      def path
+        return @path if @path
+        return nil unless orig_key
+
+        @path = @prefix + "/" + orig_key
+      end
+
+      def key
+        return @key if @key
+
+        @key = @entry[:key]
+        @key = @key[0..-3] if @key.end_with?("[]")
+        @key
+      end
+
+      # @return [LocatedEntry, nil] preceding entry that already exist in augeas
+      # tree or nil if it do not exists.
+      def preceding_existing
+        preceding_entry = preceding_tree.reverse.find do |entry|
+          entry[:operation] != :add
+        end
+
+        return nil unless preceding_entry
+
+        LocatedEntry.new(tree, preceding_entry, prefix)
+      end
+
+      # @return [true, false] returns true if there is in augeas tree any
+      # existing entry
+      def any_following?
+        following_tree.any? { |e| e[:operation] != :remove }
+      end
+
+      # @return[AugeasTree] returns augeas tree nested under entry. If there
+      # is not such tree, it create empty one.
+      def entry_tree
+        value = entry[:value]
+        case value
+        when AugeasTree then value
+        when AugeasTreeValue then value.tree
+        else AugeasTree.new
+        end
+      end
+
+      # @return [String, nil] return augeas value of entry. Can be nil.
+      # If entry value is AugeasTree, then return nil.
+      def entry_value
+        value = entry[:value]
+        case value
+        when AugeasTree then nil
+        when AugeasTreeValue then value.value
+        else value
+        end
+      end
+
+    private
+
+      # gets subtree preceding entry
+      def preceding_tree
+        tree.all_data[0..(index - 1)]
+      end
+
+      # gets subtree following entry
+      def following_tree
+        tree.all_data[(index + 1)..-1]
+      end
+
+      # index of entry in tree
+      def index
+        tree.all_data.index(entry)
+      end
+    end
+
+    # Represents operation that needs to be done after all modification.
+    # Reason to have this class is that augeas after some operations like
+    # rm or insert renumber array so previous path is no longer valid. For
+    # this reason these sensitive operations that change paths need to be done
+    # at the end and with careful order.
+    #
+    # @note This class depends on ordered operations. So adding and removing
+    # entries have to be done in order how they are placed in tree.
+    class LazyOperations
```

Comment:
This class mixes the laziness with specific augeas knowledge. In the end I would remove this class and do it all in `AugeasWriter` with  an array of procs:
```rb
@lazy_ops = []
...
@lazy_ops << ->() { add_entry(located_entry) }
@lazy_ops << ->() { aug.rm(path) }
...
@lazy_ops.reverse_each { |o| o.call }
```
but this is just a minor thing compared to all the complexity of augeas.